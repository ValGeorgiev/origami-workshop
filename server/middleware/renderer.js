import React from 'react'
import ReactDOMServer from 'react-dom/server'
import { StaticRouter } from 'react-router-dom'

// import our main App component
import App from '../../src/App';
import Navigation from '../../src/navigation';

import path from "path"
import fs from "fs"

export default (req, res, next) => {

    // point to the html file created by CRA's build tool
    const filePath = path.resolve(__dirname, '..', '..', 'build', 'index.html');

    fs.readFile(filePath, 'utf8', (err, htmlData) => {
        if (err) {
            console.error('err', err);
            return res.status(404).end()
        }

        // render the app as a string
        const html = ReactDOMServer.renderToString(
          <App origamis={req.origamis} user={req.user}>
            <StaticRouter>
              <Navigation />
            </StaticRouter>
          </App>
        );

        const htmlWithState = htmlData.replace(
          'window.__STATE__={}',
          `window.__STATE__=${JSON.stringify({
            user: req.user,
            origamis: req.origamis
          })}`
        )

        // inject the rendered app into our html and send it
        return res.send(
          htmlWithState.replace(
                '<div id="root"></div>',
                `<div id="root">${html}</div>`
            )
        );
    });
}