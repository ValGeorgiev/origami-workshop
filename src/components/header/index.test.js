import React from 'react'
import renderer from 'react-test-renderer'
import Header from '.'
import TestingEnvironment from '../../test-utils/router'
import getNavigation from '../../utils/navigation'

describe('Header Component', () => {

  it('return authenticated routes from getNavigation', () => {
    const links = getNavigation({
      loggedIn: true,
      id: '123'
    })

    const authLinks = [
      {
        title: "Publications",
        link: "/"
      },
      {
        title: "Share your thoughts",
        link: "/share"
      },
      {
        title: "Profile",
        link: `/profile/123`
      }
    ]

    expect(links).toStrictEqual(authLinks)
  })

  it('should render authenticated routes', () => {
    const tree = renderer.create(
      <TestingEnvironment value={{
        user: {
          loggedIn: true,
          id: '123'
        }
      }}>
        <Header />
      </TestingEnvironment>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render non-authenticated routes', () => {
    const tree = renderer.create(
      <TestingEnvironment value={{
        user: {
          loggedIn: false
        }
      }}>
        <Header />
      </TestingEnvironment>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})