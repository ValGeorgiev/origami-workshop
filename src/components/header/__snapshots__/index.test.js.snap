// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component should render authenticated routes 1`] = `
<header
  className="navigation"
>
  <img
    alt="logo"
    className="logo"
    src="white-origami-bird.png"
  />
  <div
    className="header-list-item"
    data-test-id="link-Publications"
  >
    <a
      className="header-link"
      href="/"
      onClick={[Function]}
    >
      Publications
    </a>
  </div>
  <div
    className="header-list-item"
    data-test-id="link-Share your thoughts"
  >
    <a
      className="header-link"
      href="/share"
      onClick={[Function]}
    >
      Share your thoughts
    </a>
  </div>
  <div
    className="header-list-item"
    data-test-id="link-Profile"
  >
    <a
      className="header-link"
      href="/profile/123"
      onClick={[Function]}
    >
      Profile
    </a>
  </div>
</header>
`;

exports[`Header Component should render non-authenticated routes 1`] = `
<header
  className="navigation"
>
  <img
    alt="logo"
    className="logo"
    src="white-origami-bird.png"
  />
  <div
    className="header-list-item"
    data-test-id="link-Publications"
  >
    <a
      className="header-link"
      href="/"
      onClick={[Function]}
    >
      Publications
    </a>
  </div>
  <div
    className="header-list-item"
    data-test-id="link-Register"
  >
    <a
      className="header-link"
      href="/register"
      onClick={[Function]}
    >
      Register
    </a>
  </div>
  <div
    className="header-list-item"
    data-test-id="link-Login"
  >
    <a
      className="header-link"
      href="/login"
      onClick={[Function]}
    >
      Login
    </a>
  </div>
</header>
`;
